libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java

```diff
@@ -348,627 +263,81 @@ public void setOnRouteSelectionChangeListener(
-
-    arrowColor = typedArray.getColor(R.styleable.NavigationMapRoute_upcomingManeuverArrowColor,
-      ContextCompat.getColor(context, R.color.mapbox_navigation_route_upcoming_maneuver_arrow_color));
-    arrowBorderColor = typedArray.getColor(R.styleable.NavigationMapRoute_upcomingManeuverArrowBorderColor,
-      ContextCompat.getColor(context, R.color.mapbox_navigation_route_upcoming_maneuver_arrow_border_color));
-
-    typedArray.recycle();
+  @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+  public void onStop() {
+    removeListeners();
   }
 
   /**
-   * Iterate through map style layers backwards till the first not-symbol layer is found.
+   * This method should be added in your {@link Activity#onDestroy()} or
+   * {@link android.support.v4.app.Fragment#onDestroyView()} to handle removing resources that were added to the map.
    */
-  private void placeRouteBelow() {
-    if (belowLayer == null || belowLayer.isEmpty()) {
-      List<Layer> styleLayers = mapboxMap.getLayers();
-      for (int i = 0; i < styleLayers.size(); i++) {
-        if (!(styleLayers.get(i) instanceof SymbolLayer)
-          // Avoid placing the route on top of the user location layer
-          && !styleLayers.get(i).getId().contains("mapbox-location")) {
-          belowLayer = styleLayers.get(i).getId();
-        }
-      }
-    }
-  }
-
-  private void drawWaypointMarkers(@NonNull MapboxMap mapboxMap, @Nullable Drawable originMarker,
-                                   @Nullable Drawable destinationMarker) {
-    if (originMarker == null || destinationMarker == null) {
-      return;
-    }
-
-    SymbolLayer waypointLayer = mapboxMap.getLayerAs(WAYPOINT_LAYER_ID);
-    if (waypointLayer == null) {
-      Bitmap bitmap = MapImageUtils.getBitmapFromDrawable(originMarker);
-      mapboxMap.addImage("originMarker", bitmap);
-      bitmap = MapImageUtils.getBitmapFromDrawable(destinationMarker);
-      mapboxMap.addImage("destinationMarker", bitmap);
-
-      waypointLayer = new SymbolLayer(WAYPOINT_LAYER_ID, WAYPOINT_SOURCE_ID).withProperties(
-        PropertyFactory.iconImage(match(
-          Expression.toString(get("waypoint")), literal("originMarker"),
-          stop("origin", literal("originMarker")),
-          stop("destination", literal("destinationMarker"))
-          )
-        ),
-        PropertyFactory.iconSize(interpolate(
-          exponential(1.5f), zoom(),
-          stop(0f, 0.6f),
-          stop(10f, 0.8f),
-          stop(12f, 1.3f),
-          stop(22f, 2.8f)
-        )),
-        PropertyFactory.iconPitchAlignment(Property.ICON_PITCH_ALIGNMENT_MAP),
-        PropertyFactory.iconAllowOverlap(true),
-        PropertyFactory.iconIgnorePlacement(true)
-      );
-      layerIds.add(WAYPOINT_LAYER_ID);
-      MapUtils.addLayerToMap(mapboxMap, waypointLayer, belowLayer);
-    }
+  @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+  public void onDestroy() {
```

Comment:
I believe this was added to workaround 👇 commented in https://github.com/mapbox/mapbox-navigation-android/pull/1387#issuecomment-444284513
> I updated the `MapRouteLine` and `MapRouteArrow` to each have an `onDestroy` that aims to clean up the layers. I also updated the arrow code to remove the layers if we find they exist.

Question (especially for @LukasPaczos) - Shouldn't this be handled gracefully from the Maps SDK side when calling `MapboxMap#onDestroy`? Have we explored this option? I can cut a ticket in `mapbox-gl-native` to discuss further if you think this should be addressed upstream (if there isn't one already 😅). Does that make sense? Let me know what you think.

Reply:
Correct, there is no need for clearing the layers on the actual `MapView` destroy. You can even see logs like
```
E/Mbgl-NativeMapView: You're calling `removeLayer` after the `MapView` was destroyed, were you invoking it after `onDestroy()`?
```
because it's called after the call to `MapView#onDestroy` and it doesn't have any real impact.

I believe that the initial idea here was to clear layers when the `NavigationMapRoute` is detached from the map so that the navigation SDK doesn't leave unused layers on the map that can still be in use otherwise. If a solution for this minor use-case seems not so relevant in the grand scheme of things, it can definitely be tackled by a separate PR.

Reply:
Sorry if I didn't make myself clear. I was wondering if the Maps SDK should be in charge of the "layers cleanup" somehow - ensuring that any layers added are removed properly when in `onDestroy`. Probably it has more implications / considerations that I'm not taking into account / missing but IMO this should be maps responsibility - obviously calling `removeLayer` afterwards is a dev implementation mistake (even in this case which doesn't seem to affect - other than the logs printed out). Is that what you're saying? What am I missing? I'm confused 😅 

> I believe that the initial idea here was to clear layers when the NavigationMapRoute is detached from the map so that the navigation SDK doesn't leave unused layers on the map that can still be in use otherwise.

Not sure if I follow, I think ☝️ was the part that make me 😕 Could you clarify?

> If a solution for this minor use-case seems not so relevant in the grand scheme of things, it can definitely be tackled by a separate PR.

Sure thing! This was a question 🤔 that I wanted to fully understand. Not a blocker at all, I agree.

Reply:
> I was wondering if the Maps SDK should be in charge of the "layers cleanup" somehow - ensuring that any layers added are removed properly when in onDestroy.

This is already the case. There is no need for a manual layer cleanup when the map is destroyed, that's what I meant above.

> Not sure if I follow, I think ☝️ was the part that make me 😕 Could you clarify?

I was going with an impression that the `NavigationMapRoute` is a standalone object that can be added and removed from the map. If that's not the case, there's nothing to worry about :)

Reply:
Yes, if the map will never outlive `MapRouteLine`, we shouldn't worry about removing those layers.

Reply:
Got it! Thanks @LukasPaczos 🙏

cc'ing @danesfeder to double check / confirm and remove.